File: src/core/python/core/model/impl/mapOD.py
from typing import List, Dict

from core.model.od import OD, ODPair


class MapOD(OD):
    """
    Class implementing the od interface with a simple map. Especially suited for sparse od matrices.
    """
    def __init__(self):
        self.od: Dict[int, Dict[int, float]] = {}

    def setValue(self, origin: int, destination: int, new_value: float) -> None:
        self.od.setdefault(origin, {})[destination] = new_value

    def computeNumberOfPassengers(self) -> float:
        return sum(value for inner_dict in self.od.values() for value in inner_dict.values())

    def getODPairs(self) -> List[ODPair]:
        return [ODPair(origin, destination, value)
                for origin in self.od.keys()
                for destination, value in self.od[origin].items()
                if value > 0]

    def getValue(self, origin: int, destination: int) -> float:
        if origin in self.od and destination in self.od[origin]:
            return self.od[origin][destination]
        raise KeyError

File: src/core/java/src/net/lintim/model/impl/MapOD.java
package net.lintim.model.impl;

import net.lintim.model.OD;
import net.lintim.model.ODPair;

import java.util.*;
import java.util.stream.Collectors;

public class MapOD implements OD {

    private Map<Integer, Map<Integer, Double>> matrix;

    public MapOD() {
        this.matrix = new HashMap<>();
    }

    @Override
    public double getValue(int origin, int destination) {
        return this.matrix.getOrDefault(origin, new HashMap<>()).getOrDefault(destination, 0.);
    }

    @Override
    public void setValue(int origin, int destination, double newValue) {
        this.matrix.putIfAbsent(origin, new HashMap<>());
        this.matrix.get(origin).put(destination, newValue);
    }

    @Override
    public double computeNumberOfPassengers() {
        return matrix.values().stream().map(m -> m.values().stream().mapToDouble(Double::doubleValue).sum())
            .mapToDouble(Double::doubleValue).sum();
    }

    @Override
    public Collection<ODPair> getODPairs() {
        return matrix.entrySet().stream()
            .flatMap(oe -> oe.getValue().entrySet().stream()
                .filter(de -> de.getValue() > 0)
                .map(de -> new ODPair(oe.getKey(), de.getKey(), de.getValue())))
            .collect(Collectors.toList());
    }
}
